modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Bigquery.scala

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2018-2021 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.storage.bigquery.streamloader
+
+import com.snowplowanalytics.snowplow.storage.bigquery.common.LoaderRow
+import com.snowplowanalytics.snowplow.storage.bigquery.common.config.model.Output
+import com.snowplowanalytics.snowplow.storage.bigquery.common.metrics.Metrics
+
+import cats.effect.{Blocker, ContextShift, Sync}
+import cats.implicits._
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.cloud.bigquery.{BigQuery, BigQueryOptions, InsertAllRequest, InsertAllResponse, TableId}
+import com.google.cloud.bigquery.InsertAllRequest.RowToInsert
+import com.permutive.pubsub.producer.PubsubProducer
+
+import scala.jdk.CollectionConverters.IterableHasAsJava
+
+object Bigquery {
+
+  final case class FailedInsert(tableRow: String) extends AnyVal
+
+  /**
+    * Insert rows into BigQuery or into "failed inserts" PubSub topic if BQ client returns
+    * any known errors.
+    *
+    * Exceptions from the underlying `insertAll` call are propagated.
+    */
+  def insert[F[_]: Sync](
+    failedInsertProducer: PubsubProducer[F, FailedInsert],
+    metrics: Metrics[F],
+    loaderRows: List[LoaderRow]
+  )(
+    mkInsert: List[LoaderRow] => F[InsertAllResponse]
+  ): F[Unit] =
+    mkInsert(loaderRows).flatMap {
+      case response if response.hasErrors =>
+        val errorIndex = response.getInsertErrors.keySet()
+        val failed     = loaderRows.zipWithIndex.filter { case (_, i) => errorIndex.contains(i.toLong) }.map(_._1)
+        val tableRows = failed.map { lr =>
+          lr.data.setFactory(new JacksonFactory)
```

Comment:
How about `JacksonFactory.getDefaultInstance` (a thread-safe singleton) instead of calling the `JacksonFactory` constructor for every item in the seq.